ml-agents/mlagents/trainers/sac/trainer.py

```diff
@@ -0,0 +1,328 @@
+        """
+        If train_interval is met, update the SAC policy given the current reward signals.
+        If reward_signal_train_interval is met, update the reward signals from the buffer.
+        """
+        if self.step % self.train_interval == 0:
+            self.trainer_metrics.start_policy_update_timer(
+                number_experiences=len(self.training_buffer.update_buffer["actions"]),
+                mean_return=float(np.mean(self.cumulative_returns_since_policy_update)),
+            )
+            self.update_sac_policy()
+            self.update_reward_signals()
+            self.trainer_metrics.end_policy_update()
+
+    def update_sac_policy(self) -> None:
+        """
+        Uses demonstration_buffer to update the policy.
+        The reward signal generators are updated using different mini batches.
+        If we want to imitate http://arxiv.org/abs/1809.02925 and similar papers, where the policy is updated
+        N times, then the reward signals are updated N times, then reward_signal_updates_per_train
+        is greater than 1 and the reward signals are not updated in parallel.
+        """
+
+        self.cumulative_returns_since_policy_update: List[float] = []
+        n_sequences = max(
+            int(self.trainer_parameters["batch_size"] / self.policy.sequence_length), 1
+        )
+
+        num_updates = self.trainer_parameters["num_update"]
+        batch_update_stats: Dict[str, list] = defaultdict(list)
+        for _ in range(num_updates):
+            LOGGER.debug("Updating SAC policy at step {}".format(self.step))
+            buffer = self.training_buffer.update_buffer
+            if (
+                len(self.training_buffer.update_buffer["actions"])
+                >= self.trainer_parameters["batch_size"]
+            ):
+                sampled_minibatch = buffer.sample_mini_batch(
+                    self.trainer_parameters["batch_size"],
+                    sequence_length=self.policy.sequence_length,
+                )
+                # Get rewards for each reward
+                for name, signal in self.policy.reward_signals.items():
+                    sampled_minibatch[
+                        "{}_rewards".format(name)
+                    ] = signal.evaluate_batch(sampled_minibatch).scaled_reward
+
+                update_stats = self.policy.update(
+                    sampled_minibatch, n_sequences, update_target=True
+                )
+                for stat_name, value in update_stats.items():
+                    batch_update_stats[stat_name].append(value)
+
+        # Truncate update buffer if neccessary. Truncate more than we need to to avoid truncating
+        # a large buffer at each update.
+        if (
+            len(self.training_buffer.update_buffer["actions"])
+            > self.trainer_parameters["buffer_size"]
+        ):
+            self.training_buffer.truncate_update_buffer(
+                int(self.trainer_parameters["buffer_size"] * 0.8)
```

Comment:
Minor -- but since `0.8` is a magic number you should probably make it a class-level constant.

Reply:
Done - also made target entropy a constant. Really should be a hyperparameter though.